cpp/include/cuspatial/experimental/hausdorff.cuh

```diff
@@ -0,0 +1,109 @@
+ * ```
+ * spaces
+ * [0 2 5] [9] [3 7]
+ *
+ * spaces represented as points per space and concatenation of all points
+ * [0 2 5 9 3 7] [3 1 2]
+ *
+ * note: the following matrices are visually separated to highlight the relationship of a pair of
+ * points with the pair of spaces from which it is produced
+ *
+ * cartesian product of all
+ * points by pair of spaces     distance between points
+ * +----------+----+-------+    +---------+---+------+
+ * : 00 02 05 : 09 : 03 07 :    : 0  2  5 : 9 : 3  7 :
+ * : 20 22 25 : 29 : 23 27 :    : 2  0  3 : 7 : 1  5 :
+ * : 50 52 55 : 59 : 53 57 :    : 5  3  0 : 4 : 2  2 :
+ * +----------+----+-------+    +---------+---+------+
+ * : 90 92 95 : 99 : 93 97 :    : 9  7  4 : 0 : 6  2 :
+ * +----------+----+-------+    +---------+---+------+
+ * : 30 32 35 : 39 : 33 37 :    : 3  1  2 : 6 : 0  4 :
+ * : 70 72 75 : 79 : 73 77 :    : 7  5  2 : 2 : 4  0 :
+ * +----------+----+-------+    +---------+---+------+
+
+ * minimum distance from
+ * every point in one           Hausdorff distance is
+ * space to any point in        the maximum of the
+ * the other space              minimum distances
+ * +----------+----+-------+    +---------+---+------+
+ * :  0       :  9 :  3    :    : 0       : 9 : 3    :
+ * :     0    :  7 :  1    :    :         :   :      :
+ * :        0 :  4 :  2    :    :         :   :      :
+ * +----------+----+-------+    +---------+---+------+
+ * :        4 :  0 :     2 :    :       4 : 0 :    2 :
+ * +----------+----+-------+    +---------+---+------+
+ * :     1    :  6 :  0    :    :         : 6 : 0    :
+ * :        2 :  2 :     0 :    :       2 :   :      :
+ * +----------+----+-------+    +---------+---+------+
+ *
+ * returned as concatenation of columns
+ * [0 2 4 3 0 2 9 6 0]
+ * ```
+ *
+ * @param[in] points_first: xs: beginning of range of (x,y) points
+ * @param[in] points_lasts: xs: end of range of (x,y) points
+ * @param[in] space_offsets_first: beginning of range of indices to each space.
+ * @param[in] space_offsets_first: end of range of indices to each space. Last index is the last
+ * @param[in] distance_first: beginning of range of output Hausdorff distance for each pair of
+ * spaces
+ *
+ * @tparam PointIt Iterator to input points. Points must be of a type that is convertible to
+ * `cuspatial::vec_2d<T>`. Must meet the requirements of [LegacyRandomAccessIterator][LinkLRAI] and
+ * be device-accessible.
+ * @tparam OffsetIt Iterator to space offsets. Value type must be integral. Must meet the
+ * requirements of [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ * @tparam OutputIt Output iterator. Must meet the requirements of
+ * [LegacyRandomAccessIterator][LinkLRAI] and be device-accessible.
+ *
+ *
+ * @pre `points_first` may equal `distance_first`, but the range `[points_first, points_last)`
+ * shall not overlap the range `[distance_first, distance_first + (points_last - points_first))
+ * otherwise.
```

Comment:
Grammar question: why "otherwise" here?

Reply:
I took this wording directly from Thrust docs. It is a little awkward. But it basically means that you can do the computation in place, but you can't have the ranges partially overlap. That said, I don't know if this makes sense in this case because the output is scalar and the input is vector, so the value_types are different sizes and therefore it can't be done in place.

Reply:
Removed.

Reply:
Removed this precondition.